mutable_batch_pb/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+            ),
+            InfluxColumnType::Field(InfluxFieldType::UInteger) => writer.write_u64(
+                &column.column_name,
+                valid_mask,
+                values.u64_values.iter().cloned(),
+            ),
+            InfluxColumnType::Tag => writer.write_tag(
+                &column.column_name,
+                valid_mask,
+                values.string_values.iter().map(|x| x.as_str()),
+            ),
+            InfluxColumnType::Field(InfluxFieldType::String) => writer.write_string(
+                &column.column_name,
+                valid_mask,
+                values.string_values.iter().map(|x| x.as_str()),
+            ),
+            InfluxColumnType::Field(InfluxFieldType::Boolean) => writer.write_bool(
+                &column.column_name,
+                valid_mask,
+                values.bool_values.iter().cloned(),
+            ),
+            InfluxColumnType::Timestamp => {
+                if valid_mask.is_some() {
+                    return Err(Error::NullTime);
+                }
+                writer.write_time(&column.column_name, values.i64_values.iter().cloned())
+            }
+            InfluxColumnType::IOx(_) => unimplemented!(),
+        }
+        .map_err(|source| Error::Write {
+            source,
+            column: column.column_name.clone(),
+        })?;
+    }
+
+    writer.commit();
+    Ok(())
+}
+
+/// Converts a potentially truncated null mask to a valid mask
+fn compute_valid_mask(null_mask: &[u8], to_insert: usize) -> Option<Vec<u8>> {
+    if null_mask.is_empty() || null_mask.iter().all(|x| *x == 0) {
+        return None;
+    }
+
+    // The expected length of the validity mask
+    let expected_len = (to_insert + 7) >> 3;
+
+    // The number of bits over the byte boundary
+    let overrun = to_insert & 7;
+
+    let mut mask: Vec<_> = (0..expected_len)
+        .map(|x| match null_mask.get(x) {
+            Some(v) => !*v,
+            None => 0xFF,
+        })
+        .collect();
+
+    if overrun != 0 {
+        *mask.last_mut().unwrap() &= (1 << overrun) - 1;
+    }
+
+    Some(mask)
+}
+
+fn pb_column_type(col: &PbColumn) -> Result<InfluxColumnType> {
+    let value_type =
+        col.values
+            .as_ref()
+            .and_then(pb_value_type)
+            .ok_or_else(|| Error::EmptyColumn {
+                column: col.column_name.clone(),
+            })?;
+
+    let semantic_type = SemanticType::from_i32(col.semantic_type);
+
+    match (semantic_type, value_type) {
+        (Some(SemanticType::Tag), InfluxFieldType::String) => Ok(InfluxColumnType::Tag),
+        (Some(SemanticType::Field), field) => Ok(InfluxColumnType::Field(field)),
+        (Some(SemanticType::Time), InfluxFieldType::Integer)
+            if col.column_name.as_str() == TIME_COLUMN_NAME =>
+        {
+            Ok(InfluxColumnType::Timestamp)
+        }
+        _ => Err(Error::InvalidType {
+            column: col.column_name.clone(),
+        }),
```

Comment:
This could be:

```suggestion
        _ => InvalidType {
            column: &col.column_name,
        }.fail(),
```

and let snafu take care of the cloning if it's needed